server/core_storage.go

```diff
@@ -488,6 +505,164 @@ DO UPDATE SET value = $6::BYTEA, version = $7, read = $8, write = $9, updated_at
+				return nil, BAD_INPUT, errors.New(fmt.Sprintf("Invalid update index %v: A client can only write their own records", i))
+			} else {
+				owner = uid.Bytes()
+			}
+		} else if caller != uuid.Nil {
+			// If the caller is a client, do not allow them to write global data.
+			return nil, BAD_INPUT, errors.New(fmt.Sprintf("Invalid update index %v: A client cannot write global records", i))
+		}
+
+		query := `
+SELECT user_id, bucket, collection, record, value, version
+FROM storage
+WHERE bucket = $1 AND collection = $2 AND user_id = $3 AND record = $4 AND deleted_at = 0`
+		if caller != uuid.Nil {
+			query += " AND write = 1"
+		}
+
+		// Query and decode the row.
+		var userID []byte
+		var bucket sql.NullString
+		var collection sql.NullString
+		var record sql.NullString
+		var value []byte
+		var version []byte
+		err = tx.QueryRow(query, update.key.Bucket, update.key.Collection, owner, update.key.Record).
+			Scan(&userID, &bucket, &collection, &record, &value, &version)
+		if err != nil && err != sql.ErrNoRows {
+			// Only fail on critical database or row scan errors.
+			// If no row was available we still allow storage updates to perform fresh inserts.
+			logger.Error("Could not update storage, query row error", zap.Error(err))
+			if e := tx.Rollback(); e != nil {
+				logger.Error("Could not update storage, rollback error", zap.Error(e))
+			}
+			return nil, RUNTIME_EXCEPTION, errors.New("Could not update storage")
+		}
+
+		// Check if we need an immediate version compare.
+		if len(update.key.Version) != 0 && !bytes.Equal(update.key.Version, version) {
+			if e := tx.Rollback(); e != nil {
+				logger.Error("Could not update storage, rollback error", zap.Error(e))
+			}
+			return nil, STORAGE_REJECTED, errors.New(fmt.Sprintf("Version check failed on update index %v: no existing value", i))
+		}
+
+		// Allow updates to create new records.
+		if len(value) == 0 {
+			value = []byte("{}")
+		}
+
+		// Apply the patch operations.
+		newValue, err := update.patch.Apply(value)
+		if err != nil {
+			if e := tx.Rollback(); e != nil {
+				logger.Error("Could not update storage, rollback error", zap.Error(e))
+			}
+			return nil, STORAGE_REJECTED, errors.New(fmt.Sprintf("Storage update index %v rejected: %v", i, err.Error()))
+		}
+		newVersion := []byte(fmt.Sprintf("%x", sha256.Sum256(newValue)))
+
+		query = `
+INSERT INTO storage (id, user_id, bucket, collection, record, value, version, read, write, created_at, updated_at, deleted_at)
+SELECT $1, $2, $3, $4, $5, $6::BYTEA, $7, $8, $9, $10, $10, 0`
+		params := []interface{}{uuid.NewV4().Bytes(), owner, update.key.Bucket, update.key.Collection, update.key.Record, newValue, newVersion, update.permissionRead, update.permissionWrite, ts}
+		if len(version) == 0 {
+			// Treat this as an if-none-match.
+			query += " WHERE NOT EXISTS (SELECT record FROM storage WHERE user_id = $2 AND bucket = $3 AND collection = $4 AND record = $5 AND deleted_at = 0)"
```

Comment:
Do we care for this? Version was already not set, meaning data was not found and it's a new record that we are storing, so no need for WHERE NOT EXISTS?

-- 

Also, for an actual "if-none-match" case, shall we not check if version == "*" ?